rust/kernel/sync/guard.rs

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! A generic lock guard and trait.
+//!
+//! This module contains a lock guard that can be used with any locking primitive that implements
+//! the ([`Lock`]) trait. It also contains the definition of the trait, which can be leveraged by
+//! other constructs to work on generic locking primitives.
+
+/// Allows mutual exclusion primitives that implement the [`Lock`] trait to automatically unlock
+/// when a guard goes out of scope. It also provides a safe and convenient way to access the data
+/// protected by the lock.
+#[must_use = "the lock unlocks immediately when the guard is unused"]
+pub struct Guard<'a, L: Lock + ?Sized> {
+    pub(crate) lock: &'a L,
+}
+
+// SAFETY: `Guard` is sync when the data protected by the lock is also sync. This is more
+// conservative than the default compiler implementation; more details can be found on
+// https://github.com/rust-lang/rust/issues/41622 -- it refers to `MutexGuard` from the standard
+// library.
+unsafe impl<L> Sync for Guard<'_, L>
+where
+    L: Lock + ?Sized,
+    L::Inner: Sync,
+{
+}
+
+impl<L: Lock + ?Sized> core::ops::Deref for Guard<'_, L> {
+    type Target = L::Inner;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: The caller owns the lock, so it is safe to deref the protected data.
+        unsafe { &*self.lock.locked_data().get() }
+    }
+}
+
+impl<L: Lock + ?Sized> core::ops::DerefMut for Guard<'_, L> {
+    fn deref_mut(&mut self) -> &mut L::Inner {
+        // SAFETY: The caller owns the lock, so it is safe to deref the protected data.
+        unsafe { &mut *self.lock.locked_data().get() }
+    }
+}
+
+impl<L: Lock + ?Sized> Drop for Guard<'_, L> {
+    fn drop(&mut self) {
+        // SAFETY: The caller owns the lock, so it is safe to unlock it.
+        unsafe { self.lock.unlock() };
+    }
+}
+
+impl<'a, L: Lock + ?Sized> Guard<'a, L> {
+    /// Constructs a new lock guard.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that it owns the lock.
+    pub(crate) unsafe fn new(lock: &'a L) -> Self {
+        Self { lock }
+    }
+}
+
+/// A generic mutual exclusion primitive.
+///
+/// [`Guard`] is written such that any mutual exclusion primitive that can implement this trait can
+/// also benefit from having an automatic way to unlock itself.
+pub trait Lock {
+    /// The type of the data protected by the lock.
+    type Inner: ?Sized;
+
+    /// Acquires the lock, making the caller its owner.
+    fn lock_noguard(&self);
```

Comment:
Hi, late for the party. But I wonder why this is not `unsafe`, because (put aside recursive locks) this must be called when the current thread is not the owner of the lock?

Reply:
Because it doesn't affect memory safety.

The following rust program, using the standard library, compiles fine (but of course deadlocks):
```
use std::sync::Mutex;

fn main() {
    let x = Mutex::new(0);
    let g = x.lock();
    let h = x.lock();
    *g.unwrap() = 10;
    *h.unwrap() = 20;
}
```

From https://doc.rust-lang.org/nomicon/what-unsafe-does.html:
> Rust is otherwise quite permissive with respect to other dubious operations. Rust considers it "safe" to:
>
> * Deadlock

Reply:
So `unlock` and `locked_data` are `unsafe`, because calling them without the lock being held may access the memory (lock or data) which is not owned by current thread, and that breaks memory safety, right?

Reply:
That's correct.

One thing that I noticed over the weekend is that `locked_data` is actually safe because it only returns a reference to `UnsafeCell` -- the actual unsafety is when accessing the data, which already requires an unsafe block. So I am removing the `unsafe` tag from `locked_data` in my next pull request (for `LockedBy`) which I intend upload later today.

Reply:
Thanks for the explanation ;-) . Indeed, returning a reference to `UnsafeCell` is safe because a raw pointer dereference (which requires `unsafe`) is required to access the memory.

Reply:
No problem. Keep the questions coming when things don't seem to make sense!

We're always happy to have more people reviewing and contributing :)

Reply:
We are all learning here, so indeed, any comments, questions and contributions are very welcome :)

And any & all APIs are nevertheless in flux/experimental -- it will take some time until we figure out how to model best for all the different things the kernel offers.